### PR TITLE
Resource operators for multi qubit operations

### PR DESCRIPTION
**Pull request by willjmax**
[sc-76761]

### CODE REVIEW

File: pennylane/labs/resource_estimation/__init__.py [hunk 1]
```diff
@@ -61,16 +61,30 @@
...
     ~ResourceIdentity
+    ~ResourceIsingXX
+    ~ResourceIsingXY
+    ~ResourceIsingYY
+    ~ResourceIsingZZ
+    ~ResourceMultiRZ
+    ~ResourceOrbitalRotation
     ~ResourceMultiControlledX
     ~ResourcePhaseShift
+    ~ResourcePSWAP
```
Comment (Jaybsoni) [inline pennylane/labs/resource_estimation/__init__.py:79]: Missing `ResourcePauliRot`

File: pennylane/labs/resource_estimation/ops/qubit/parametric_ops_multi_qubit.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+# limitations under the License.
+r"""Resource operators for parametric multi qubit operations."""
+import pennylane as qml
+import pennylane.labs.resource_estimation as re
+
+# pylint: disable=arguments-differ
+
+
+class ResourceMultiRZ(qml.MultiRZ, re.ResourceOperator):
+    r"""Resource class for the MultiRZ gate.
```
Comment (Jaybsoni) [inline pennylane/labs/resource_estimation/ops/qubit/parametric_ops_multi_qubit.py:22]: Missing reference for the resources method.

File: pennylane/labs/resource_estimation/ops/qubit/parametric_ops_multi_qubit.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+    def _resource_decomp(pauli_word, **kwargs):
+        if set(pauli_word) == {"I"}:
+            gp = re.ResourceGlobalPhase.resource_rep(**kwargs)
+            return {gp: 1}
+
+        active_wires = len(pauli_word.replace("I", ""))
+
+        h = re.ResourceHadamard.resource_rep(**kwargs)
+        rx = re.ResourceRX.resource_rep(**kwargs)
+        multi_rz = re.ResourceMultiRZ.resource_rep(active_wires, **kwargs)
```
Comment (Jaybsoni) [inline pennylane/labs/resource_estimation/ops/qubit/parametric_ops_multi_qubit.py:68]: Actually, I would suggest going straight to CNOT and 1 Z here. No need to go through the MultiRZ. This is one of those cases where we will get a significant performance boost for doing a single step simplification without losing much for generality.

File: pennylane/labs/resource_estimation/ops/qubit/qchem_ops.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+                    0 & \cos(\phi/2) & -\sin(\phi/2) & 0 \\
+                    0 & \sin(\phi/2) & \cos(\phi/2) & 0 \\
+                    0 & 0 & 0 & 1
+                \end{bmatrix}.
+
+    """
+
+    @staticmethod
+    def _resource_decomp(*args, **kwargs):
+        """TODO: implement in resource_symbolic_ops branch"""
```
Comment (Jaybsoni) [inline pennylane/labs/resource_estimation/ops/qubit/qchem_ops.py:36]: What is required here?

File: pennylane/labs/resource_estimation/ops/qubit/qchem_ops.py [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+        .. math::
+
+            &|0011\rangle \rightarrow \cos(\phi/2) |0011\rangle + \sin(\phi/2) |1100\rangle\\
+            &|1100\rangle \rightarrow \cos(\phi/2) |1100\rangle - \sin(\phi/2) |0011\rangle,
+    """
+
+    @staticmethod
+    def _resource_decomp(*args, **kwargs):
+        """See https://arxiv.org/abs/2104.05695"""
```
Comment (Jaybsoni) [inline pennylane/labs/resource_estimation/ops/qubit/qchem_ops.py:135]: Link this above.

File: pennylane/labs/resource_estimation/ops/qubit/qchem_ops.py [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+        .. math::
+
+            &|0011\rangle \rightarrow \cos(\phi/2) |0011\rangle - \sin(\phi/2) |1100\rangle\\
+            &|1100\rangle \rightarrow \cos(\phi/2) |1100\rangle + \sin(\phi/2) |0011\rangle\\
+            &|x\rangle \rightarrow e^{-i\phi/2} |x\rangle,
+    """
+
+    @staticmethod
+    def _resource_decomp(*args, **kwargs):
+        """TODO: implement in resource_symbolic_op branch"""
```
Comment (Jaybsoni) [inline pennylane/labs/resource_estimation/ops/qubit/qchem_ops.py:171]: What's missing?

**Review by Jaybsoni — COMMENTED** [review]
Just a few comments to wrap up this PR: 

1. The "Resources" section for each of the operation should at least have enough information that a user can know what the expected resources are without having to run the code. In most cases we should have a paper with either the resource decomposition or a numeric expression for the cost. In the case that the gate is very fundamental, you should list out the resources or circuit diagram (see the docs for `ResourceToffoli` as an example). 
2. Regarding the QChem ops tests. Even though the implementation is still TODO, you should still test that the resources raise a NotImplemented error and that the other methods are as expected. In general, if a feature is TODO, we either don't add any code for it (just the comment) or if we add code that's empty, we make sure to test it produces empty results. 

3. Missing changelog entry. See my PRs for how to add a changelog entry. 

Once those are addressed and the docs build I am happy to approve this 👍🏼 Thanks

File: pennylane/labs/resource_estimation/ops/qubit/parametric_ops_multi_qubit.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        return re.CompressedResourceOp(cls, {"num_wires": num_wires})
+
+
+class ResourcePauliRot(qml.PauliRot, re.ResourceOperator):
+    r"""Resource class for the PauliRot gate.
+
+    Resources:
+        The resources come from Section VIII of "The Bravyi-Kitaev transformation for quantum computation
+        of electronic structure" (https://arxiv.org/pdf/1208.5986). See Figure 4 of that section for
+        an illustration.
```
Comment (Jaybsoni) [inline pennylane/labs/resource_estimation/ops/qubit/parametric_ops_multi_qubit.py:55]:
```suggestion
    Resources:
        The resources come from Section VIII (figures 3, 4) of `The Bravyi-Kitaev transformation for 
        quantum computation of electronic structure <https://arxiv.org/pdf/1208.5986>`_ paper, in 
        combination with the following identity:

        .. math::

            \begin{align}
                \hat{X} &= \hat{H} \cdot \hat{Z} \cdot \hat{H}, \\
                \hat{Y} &= \hat{S} \cdot \hat{H} \cdot \hat{Z} \cdot \hat{H} \cdot \hat{S}^{\dagger}.
            \end{align}

        Specifically, the resources are given by one :code:`RZ` gate and a cascade of :math:`2 * (n - 1)`
        :code:`CNOT` gates where :math:`n` is the number of qubits the gate acts on. Additionally, for 
        each :code:`X` gate in the Pauli word we conjugate by a :code:`Hadamard` gate, and for each
        :code:`Y` gate in the Pauli word we conjugate by :code:`Hadamard` and :code:`S` gates.

```

File: pennylane/labs/resource_estimation/ops/qubit/parametric_ops_multi_qubit.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+        .. math:: XX(\phi) = \exp\left(-i \frac{\phi}{2} (X \otimes X)\right) =
+            \begin{bmatrix} =
+                \cos(\phi / 2) & 0 & 0 & -i \sin(\phi / 2) \\
+                0 & \cos(\phi / 2) & -i \sin(\phi / 2) & 0 \\
+                0 & -i \sin(\phi / 2) & \cos(\phi / 2) & 0 \\
+                -i \sin(\phi / 2) & 0 & 0 & \cos(\phi / 2)
+            \end{bmatrix}.
+
+        The circuit implementing this transformation is given by
```
Comment (Jaybsoni) [inline pennylane/labs/resource_estimation/ops/qubit/parametric_ops_multi_qubit.py:112]:
```suggestion
        The circuit implementing this transformation is given by:
```

**Review by Jaybsoni — CHANGES_REQUESTED** [review]
Some suggestions about the docstrings and one more efficient resources method.

**Review by Jaybsoni — APPROVED** [review]
Looks good to me 👍🏼

**Review by soranjh — APPROVED** [review]

**Pull request merged**